### PR TITLE
Don't require a stack to run `pulumi query`

### PR DESCRIPTION
**Pull request by hausdorff**
~Not entirely ready for review, but the basic idea is to remove the requirement that `pulumi query` run inside a Pulumi stack. Now, all that should be necessary is the ability to log into the backend (currently only supporting the cloud backend).~

The reason this is possible is because `pulumi query` doesn't actually need access to a default stack, necessarily—the query primitives are implemented purely inside the SDK, so it needs only to be able to _get_ a snapshot of a stack, feed it into the `query.from`, and query the result.

A outgrowth of this is that `pulumi query` now actually doesn't support running inside a stack at all—this is actually intentional. If you want to use query primitives inside a stack, you should run `pulumi up` or `pulumi preview`. Unfortunately, this is not really enforced yet, so if you do run `pulumi query` in the context of a stack, you'll get some mysterious errors when you run `getStack` because the name will be empty.

### CODE REVIEW

**Comment by hausdorff** [comment]
@pgavlin I'm starting to get closer to all the other pieces we need here—provider querying, streaming invoke. In the meantime, I think we should provisionally aim to get this through in preparation? cc @lukehoban

**Comment by hausdorff** [comment]
@pgavlin I am going to put the almost-ready WIP commit that wires up provider querying while I'm here. I'll fix it up in a bit, but it should give you a sense of where this is going.

File: pkg/resource/deploy/source_query.go [hunk 1]
```diff
@@ -168,15 +198,61 @@ func runLangPlugin(src *querySource) result.Result {
...
+	}
+
+	go func() {
+		for {
+			select {
+			case e := <-regChan:
+				urn := resource.NewURN(
+					"query-stack", "query-project", "parent-type", e.goal.Type, e.goal.Name)
+
+				inputs, _, err := reg.Check(urn, resource.PropertyMap{}, e.goal.Properties, false)
```
Comment (hausdorff) [inline pkg/resource/deploy/source_query.go]: We should probably include `CheckFailures` in the result here, but I'm still working through how to flow that back through the results.

File: pkg/resource/deploy/source_query.go [hunk 2]
```diff
@@ -168,15 +198,61 @@ func runLangPlugin(src *querySource) result.Result {
...
+		regChan:         regChan,
+		cancel:          cancel,
+	}
+
+	go func() {
+		for {
+			select {
+			case e := <-regChan:
+				urn := resource.NewURN(
+					"query-stack", "query-project", "parent-type", e.goal.Type, e.goal.Name)
```
Comment (hausdorff) [inline pkg/resource/deploy/source_query.go]: This totally sucks, but I don't _think_ we have much precedent for generating a totally fake stack, fake project, and fake parent for a resource type? It seems worse to change how the `Check` interface handles URNs generally, which I think implies that we actually do want to generate this. But something about it just feels kind of wrong. Would love to hear what you think, @pgavlin.

**Comment by hausdorff** [comment]
@pgavlin ready for first round of comments, I think.

The most "interesting" part of this PR is the last part, where we expose the resource providers to `query`, which will eventually allow us to use a resource provider's `Invoke` to "run queries" (_e.g._, in the case of Kubernetes, to list all resources of some type).

This is "interesting" principally because of how errors are propagated back to the "engine" (in scare quotes here because it's basically just a naked `querySource`—no `StepExecutor` or the like). In traditional Pulumi programs, we depend on the execution primitives to register resource providers and propagate errors back gracefully. But because we don't have them, these have to be reported back by the `querySource` itself.

The reasons for this are primarily because (as I understand it) `Invoke` on a resource provider will trigger it to load from `defaultProviders` if it does not already exist. That implies that the (mostly-noop) `queryResmon` has to know about the `defaultProviders` and broker the requests to invoke to point at those `defaultProviders`, which in turn implies that it must keep track of which of those provider loads have failed.

For this reason, we put a _very_ lightweight version of this error handling logic directly into the `querySource` and `queryResmon`, and simply report that back on `querySource#Wait()` when it occurs. This means minimal changes on the "engine" side, which again is simply invoking the `querySource`.

File: pkg/engine/events.go [hunk 3]
```diff
@@ -169,7 +169,7 @@ type StepEventStateMetadata struct {
 func makeEventEmitter(events chan<- Event, update UpdateInfo) (eventEmitter, error) {
 	target := update.GetTarget()
 	var secrets []string
-	if target.Config.HasSecureValue() {
+	if target != nil && target.Config.HasSecureValue() {
```
Comment (pgavlin) [inline pkg/engine/events.go]: Is this change necessary?

File: pkg/resource/deploy/source_query.go [hunk 4]
```diff
@@ -99,10 +116,15 @@ func (src *querySource) Wait() result.Result {
...
 
 	select {
-	case src.res = <-src.finChan:
+	case src.res = <-src.langPluginFinChan:
 		// Language plugin has exited. No need to call `Close`.
 		src.done = true
 		return src.res
+	case src.res = <-src.providerRegErrChan:
+		// Provider registration has failed.
+		src.done = true
```
Comment (pgavlin) [inline pkg/resource/deploy/source_query.go]: Nit: this isn't necessary--the call to `src.Close()` will cover this.

File: pkg/resource/deploy/source_query.go [hunk 5]
```diff
@@ -168,15 +198,60 @@ func runLangPlugin(src *querySource) result.Result {
...
+		requests:        make(chan defaultProviderRequest),
+		regChan:         regChan,
+		cancel:          cancel,
+	}
+
+	go func() {
+		for {
+			select {
+			case e := <-regChan:
+				urn := syntheticURN(e.goal)
```
Comment (pgavlin) [inline pkg/resource/deploy/source_query.go]: This is specifically for registering providers?

File: pkg/resource/deploy/source_query.go [hunk 6]
```diff
@@ -289,13 +373,66 @@ func (rm *queryResmon) SupportsFeature(ctx context.Context,
...
-	case "secrets":
-		hasSupport = true
-	}
-
 	return &pulumirpc.SupportsFeatureResponse{
 		HasSupport: hasSupport,
 	}, nil
 }
+
+func (rm *queryResmon) parseProviderRequest(pkg tokens.Package, version string) (providers.ProviderRequest, error) {
```
Comment (pgavlin) [inline pkg/resource/deploy/source_query.go]: Are these copied from the normal resource monitor? If so, can we refactor them to share the code?

File: pkg/resource/deploy/source_query.go (hunk 5 above)
Comment (hausdorff) [reply pkg/resource/deploy/source_query.go]: It is.

File: pkg/resource/deploy/source_query.go (hunk 6 above)
Comment (hausdorff) [reply pkg/resource/deploy/source_query.go]: We can.

File: pkg/engine/events.go (hunk 3 above)
Comment (hausdorff) [reply pkg/engine/events.go]: Yeah—I forget where, but somewhere in the engine code this value is interrogated and it crashes `query`.

File: pkg/resource/deploy/source_query.go (hunk 4 above)
Comment (hausdorff) [reply pkg/resource/deploy/source_query.go]: Let's remove it from the case below, too, then.

File: pkg/resource/deploy/source_query.go (hunk 5 above)
Comment (hausdorff) [reply pkg/resource/deploy/source_query.go]: Renamed.

File: pkg/engine/events.go (hunk 3 above)
Comment (pgavlin) [reply pkg/engine/events.go]: I thought we used `makeQueryEventEmitter` for query mode?

File: pkg/resource/deploy/source_query.go [hunk 7]
```diff
@@ -168,15 +196,57 @@ func runLangPlugin(src *querySource) result.Result {
...
+func newQueryResourceMonitor(
+	builtins *builtinProvider, defaultProviderVersions map[tokens.Package]*semver.Version,
+	provs ProviderSource, reg *providers.Registry, plugctx *plugin.Context,
+	providerRegErrChan chan<- result.Result, tracingSpan opentracing.Span) (*queryResmon, error) {
 
 	// Create our cancellation channel.
 	cancel := make(chan bool)
 
+	// Create channel for handling registrations.
+	regChan := make(chan *registerResourceEvent)
```
Comment (pgavlin) [inline pkg/resource/deploy/source_query.go]:
```suggestion
	providerRegChan := make(chan *registerResourceEvent)
```
